syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.github.yzernik.squeaknode";
option java_outer_classname = "SqueakProto";
option objc_class_prefix = "SQK";

package squeaknode;

// Interface exported by the server.
service SqueakServer {

  /** sqk: `postsqueak`
  */
  rpc UploadSqueak (UploadSqueakRequest) returns (UploadSqueakReply) {}

  /** sqk: `getsqueak`
  */
  rpc DownloadSqueak (DownloadSqueakRequest) returns (DownloadSqueakReply) {}

  /** sqk: `lookupsqueakstodownload`
  */
  rpc LookupSqueaksToDownload (LookupSqueaksToDownloadRequest) returns (LookupSqueaksToDownloadReply) {}

  /** sqk: `lookupsqueakstoupload`
  */
  rpc LookupSqueaksToUpload (LookupSqueaksToUploadRequest) returns (LookupSqueaksToUploadReply) {}

  /** sqk: `getoffer`
  */
  rpc GetOffer (GetOfferRequest) returns (GetOfferReply) {}

}

message UploadSqueakRequest {
    /// Content of the squeak to be made.
    Squeak squeak = 1;
}

message DownloadSqueakRequest {
    /// Hash of the squeak to get.
    bytes hash = 1;
}

message LookupSqueaksToDownloadRequest {
    /// List of the author addresses
    repeated string addresses = 1;
    /// Minimum block to look for
    int32 min_block = 2;
    /// Maximum block to look for
    int32 max_block = 3;
}

message LookupSqueaksToUploadRequest {
    /// List of the author addresses
    repeated string addresses = 1;
}

message GetOfferRequest {
    /// Hash of the squeak to buy.
    bytes hash = 1;
}

message UploadSqueakReply {
}

message DownloadSqueakReply {
    /// The squeak.
    Squeak squeak = 1;
}

message LookupSqueaksToDownloadReply {
    /// Hash of the squeak to get.
    repeated bytes hashes = 1;
}

message LookupSqueaksToUploadReply {
    /// Hash of the squeak to get.
    repeated bytes hashes = 1;
    /// List of the author addresses
    repeated string addresses = 2;
    /// Minimum block to look for
    int32 min_block = 3;
    /// Maximum block to look for
    int32 max_block = 4;
}

message GetOfferReply {
    /// The offer
    Offer offer = 1;
}

message Squeak {
    /// Hash of the squeak.
    bytes hash = 1;

    /// Serialized squeak.
    bytes serialized_squeak = 2;
}

message Offer {
    /// The squeak hash.
    bytes squeak_hash = 1;

    /// Nonce
    bytes nonce = 2;

    /// The invoice
    string payment_request = 3;

    /// The host of the seller lightning node
    string host = 4;

    /// The port of the seller lightning node
    int32 port = 5;
}

// message Interested {
//     /// List of the author addresses
//     repeated string addresses = 1;
//     /// Minimum block to look for
//     int32 min_block = 2;
//     /// Maximum block to look for
//     int32 max_block = 3;
// }
